solidity/contracts/token-distribution/TokenDistributor.sol

```diff
@@ -0,0 +1,145 @@
+pragma solidity 0.5.17;
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/SafeERC20.sol";
+import "openzeppelin-solidity/contracts/cryptography/MerkleProof.sol";
+
+/// @title Token Distributor
+/// @notice This contract can be used to distribute ERC20 tokens
+/// @dev This contract is based on the Uniswap's Merkle Distributor
+/// https://github.com/Uniswap/merkle-distributor with some modifications:
+/// - added 'allocate()' function that will be called to allocate tokens for a
+///   merkle root,
+/// - added a possibility for recipient to redirect tokens withdrawal to another
+///   address by providing a signature over that address.
+contract TokenDistributor is Ownable {
+    using SafeERC20 for IERC20;
+
+    IERC20 public token;
+
+    bytes32 public merkleRoot;
+
+    mapping(uint256 => uint256) private claimedBitMap;
+
+    event TokensAllocated(bytes32 merkleRoot, uint256 amount);
+
+    event TokensClaimed(
+        uint256 indexed index,
+        address indexed recipient,
+        address indexed destination,
+        uint256 amount
+    );
+
+    constructor(IERC20 _token) public {
```

Comment:
Not sure 100%, but if we pass an address of _token (cKEEP), wouldn't it simplify a deployment script a bit?

Reply:
I wanted this contract to be generic and usable with any ERC20.

Reply:
Sure, but I think if you pass an address `constructor(address _token)` it still stays generic. When we'll have a Token address (being it cKEEP or other token) we can just pass an address of the deployed token here, right? I guess it won't simplify things in this contract, but just asking about the deployment script whether it would be easier passing an address of already deployed token instead of initiating it first and then passing here. Not biggy, just wanted to ask.